### DISCUSSION TITLE
Issues reliably creating mesh from point cloud

### DISCUSSION THREAD

**Discussion by slartz42** [pyvista/pyvista#3828] [Q&A]
Hello,

I'm currently using a extrude_rotate on a 2D point cloud to create a visual representation of a 3D model, it working quite well. Here is a picture and VTK file of the 2D point cloud I'm using:

![2D Point Cloud](https://user-images.githubusercontent.com/52103361/212549761-38ced64b-f26b-4bfc-8696-a3e2896fe059.PNG)

[2D_point_cloud.zip](https://github.com/pyvista/pyvista/files/10420515/2D_point_cloud.zip)

Here is the 3D point cloud after extrude_rotate:

![3D Point Cloud](https://user-images.githubusercontent.com/52103361/212549727-7f505868-1461-4e65-b27a-77ca99deccb3.PNG)

[3D_point_cloud.zip](https://github.com/pyvista/pyvista/files/10420517/3D_point_cloud.zip)

The issue is when I try to create mesh using reconstruct_surface, the 3D model gets all wonky. I've tried using just about every function / parameter available in the API to fix this, but I can't seem to fix it. I'm not sure if the issue is due to the model being hollow or not

![Mesh Example](https://user-images.githubusercontent.com/52103361/212549970-b2173e7f-067b-4458-8302-c5eeebe23cd0.PNG)

I've also tried converting the 2D point cloud to a mesh before revolving, but Delauney_2D keeps making triangles outside of my profile. I tried using the edge_source parameter using lines for workpiece boundary, but this still did not work:

![Delauney 2D](https://user-images.githubusercontent.com/52103361/212550082-4d0de77a-0356-4669-96ab-30f853069d3a.PNG)

![2D Lines](https://user-images.githubusercontent.com/52103361/212550098-d57c888c-23a6-4860-83e2-ce13c9e2521f.PNG)

There is probably something I am missing here, any help would be greatly appreciated!

**Comment by banesullivan** [comment]
This is a very dense point cloud and indeed, I'm seeing issues running any of our typical surface reconstruction methods with this... 

I'm trying out some code from Vedo (with some modifications to speed up and work with PyVista) to see what I can make...

vedo actually has a nice implementation of `vtkExtractSurface`: https://github.com/marcomusy/vedo/blob/85e2e2ba9a6e40d001ac4145c2787aab8b246fe1/vtkplotter/analysis.py#L1212

We should add an this in PyVista regardless, but I'm working to see if this will produce nice results for your point cloud without taking ages to execute...

<details>

```py
def reco_surface(points, bins=256):
    """
    Originally from Vedo but modified.
    """

    if isinstance(points, vtk.vtkActor):
        points = points.coordinates()
    N = len(points)
    if N < 50:
        print("recoSurface: Use at least 50 points.")
        return None
    points = np.array(points)

    ptsSource = vtk.vtkPointSource()
    ptsSource.SetNumberOfPoints(N)
    ptsSource.Update()
    vpts = ptsSource.GetOutput().GetPoints()
    # for i, p in enumerate(points):
    #     vpts.SetPoint(i, p)
    
    ptsSource.GetOutput().SetPoints(pv.vtk_points(points))
    polyData = ptsSource.GetOutput()

    distance = vtk.vtkSignedDistance()
    f = 0.1
    x0, x1, y0, y1, z0, z1 = polyData.GetBounds()
    distance.SetBounds(x0-(x1-x0)*f, x1+(x1-x0)*f,
                       y0-(y1-y0)*f, y1+(y1-y0)*f,
                       z0-(z1-z0)*f, z1+(z1-z0)*f)
    if polyData.GetPointData().GetNormals():
        distance.SetInputData(polyData)
    else:
        normals = vtk.vtkPCANormalEstimation()
        normals.SetInputData(polyData)
        normals.SetSampleSize(int(N / 50))
        normals.SetNormalOrientationToGraphTraversal()
        distance.SetInputConnection(normals.GetOutputPort())
        print("Recalculating normals for", N, "Points, sample size=", int(N / 50))

    b = polyData.GetBounds()
    diagsize = np.sqrt((b[1] - b[0]) ** 2 + (b[3] - b[2]) ** 2 + (b[5] - b[4]) ** 2)
    radius = diagsize / bins * 5
    distance.SetRadius(radius)
    distance.SetDimensions(bins, bins, bins)
    distance.Update()

    print("Calculating mesh from points with R =", radius)
    surface = vtk.vtkExtractSurface()
    surface.SetRadius(radius * 0.99)
    surface.HoleFillingOn()
    surface.ComputeNormalsOff()
    surface.ComputeGradientsOff()
    surface.SetInputConnection(distance.GetOutputPort())
    surface.Update()
    return surface.GetOutput()
```

</details>

**Reply by slartz42** [reply]
I'm trying another route that I think will be better to ensure the whole 3D object is meshed instead of just the surfaces, I just created a mesh grid using numpy and remove points that fell outside of my profile.

The problem I am still having is with Delaunay_2d connecting lines outside of my profile, do you know what might be causing this?

![image](https://user-images.githubusercontent.com/52103361/214862276-6cad1ceb-cd0b-40d3-83a6-e603cb15a766.png)

Here is the point cloud before Delaunay_2d:

![image](https://user-images.githubusercontent.com/52103361/214864673-59ed6bd3-a082-4afb-9d81-abd55a3b4838.png)

Here are the lines I am using as the edge_source parameter:

![image](https://user-images.githubusercontent.com/52103361/214864512-4718200d-b333-425c-bd27-8545096e2924.png)

**Reply by slartz42** [reply]
Hello, just wanted to see if you had any suggestions, I'm still trying to work through this issue. Mainly trying to constrain Delaunay 2D to the correct profile shape, would it be easier to just extrude_revolve and then use extrude_trim to remove the extra lines?

**Reply by slartz42** [reply]
Nevermind I got it figured out, I was able to use the alpha parameter to get it to work

**Comment by slartz42** [comment]
Thanks for the help, hopefully it works. I can probably reduce the resolution of the extrude_revolve function to make it easier to work with if necessary
